### DISCUSSION TITLE
Release of QCheck2

### DISCUSSION THREAD

**Discussion by vch9** [c-cube/qcheck#134] [General]
Hi!
Do you have any idea on when you'll release the new QCheck2?
@Sir4ur0n and I are currently working on migrating the Tezos qcheck tests to qcheck2 -> [here](https://gitlab.com/tezos/tezos/-/merge_requests/2908).
That merge request should be a good example of the the "validity" of that new version.

(I'm not trying to push you, we were just wondering :smile:)

**Comment by c-cube** [comment]
That's a good question. The only PR pending is a stylistic one, I think we could give a try to a release with qcheck2. The stability guarantees are going to be weak for a little while though :slightly_smiling_face:.

**Reply by vch9** [reply]
From a user perspective, #107 and/or #102 could be very useful. I don't really know if someone would work on this in the following days/weeks (idk for myself).
Otherwise the MR on Tezos and future new property based tests could reassure us about qcheck guarantees.

**Reply by c-cube** [reply]
I've kicked off a release anyway, but don't worry, QCheck2 won't change _drastically_. Small things might change but the hard stuff (generators, etc.) won't break much I think.

**Reply by jmid** [reply]
I just gave the new QCheck2 a quick spin - and unfortunately found regressions in the first two properties I tested.
Perhaps we want to hold off a release until it has been tested a bit more?

**Reply by c-cube** [reply]
indeed, it's a draft PR for now. I've been re-adding some functions that were removed from `QCheck` and that were found by opam's extensive testing.

**Reply by c-cube** [reply]
@jmid would you mind submitting these properties, if they're small enough, so we can build some regression tests?

also, did anything based on `QCheck` (not `QCheck2`) break?

**Reply by jmid** [reply]
> @jmid would you mind submitting these properties, if they're small enough, so we can build some regression tests?

OK, can submit them both as `QCheck(1)`+`QCheck2` tests.
I can see a file test/core/test.ml containing a number of newer `QCheck2` tests.
I could have sworn we had a collection of old regression tests (it may not have been using `OUnit(2)`)
just a file with a bunch of "run on these before release" in old versions 0.3, 0.4, 0.5 or so.
IIRC @c-cube, didn't we use some of my playing-with-the-function-generator https://github.com/jmid/qcheck-fun
as regression tests?

> also, did anything based on QCheck (not QCheck2) break?

Good point, indeed something could have been broken as part of the unification.
I can check some of my prior projects if they still work.

**Reply by c-cube** [reply]
>OK, can submit them both as QCheck(1)+QCheck2 tests.

Awesome, please do!

> IIRC @c-cube, didn't we use some of my playing-with-the-function-generator https://github.com/jmid/qcheck-fun
as regression tests?

I don't recall that we used these as regression tests, but I'd welcome a PR that adds them to the repo! Never sure how to integrate them into actual automated tests (maybe nested tests?) but if we could put that into CI it'd be awesome.

**Reply by jmid** [reply]
OK, I've had a closer look. From what I can tell
- `src/core/example/QCheck_runner_test.ml` contains a combination of examples and tests (incl. at least one regression and some of the function tests from the above link) - runnable by `make example-runner`
- `src/core/example/alcotest` and `src/core/example/ounit` contain selected variations of these - runnable by `make example-test`, `make example-ounit-test`, and `make example-alcotest`
- `test/core/test.ml` contains the beginning of a proper test suite, starting with a bunch of internal QCheck2 tests from @Sir4ur0n - runnable by `make test`

The last one fits immediately into a CI hook, whereas the first two have good content accumulated over a couple of years, but morally more of the form "run-these-and-manually-check-the-output-errors-before-releasing, please".
I believe this is because the former are all positive tests - they should be "all green".

To turn the latter into the former - and to add the two false examples above as proper testsuite tests I think
we could benefit from a `fails`-test. This should test a knowingly false property and ensure that it fails as expected.
The commercial Quviq QuickCheck for Erlang has such a thing: http://quviq.com/documentation/eqc/eqc.html#fails-1
Do you agree? (A quick look at the source code indicates that this may be achievable by adding a `FalseTest` case to `Test.t` + a number of adjustments)

**Reply by c-cube** [reply]
actually, now that I think of it: we could turn these into _expect_ tests. This allows to check for the output shape as well as pass/fail status, and is very low cost to implement.

**Reply by c-cube** [reply]
fyi: https://github.com/c-cube/qcheck/pull/140

**Reply by jmid** [reply]
Awesome @c-cube! I've come across expect-tests in the past, but hadn't thought of using them here...
I agree this is more low cost to implement and I'm glad to see our accumulated tests come to good use! 😀👍🏻

**Comment by c-cube** [comment]
Closed the release PR to opam-repo as the new code is clearly not cooked yet :)

**Comment by c-cube** [comment]
Indeed! I just merged it. It's a bit finicky because some output is not deterministic, but hopefully it serves us well.

**Reply by jmid** [reply]
Great! Can I ask why you only diff-test for OCaml pre-4.09? (just curious) 🤔

**Comment by c-cube** [comment]
On 4.12 some compiler output changed again. But now it's filtered out, maybe I should retry on 4.12

**Comment by sir4ur0n** [comment]
Hey, what's the current status on releasing QCheck2 please? Now that the String shrinking bug is fixed.

I have rebased https://gitlab.com/tezos/tezos/-/merge_requests/2908 on `master` and adapted to latest changes, tests still pass - the only noticeable difference is for `src/lib_proxy/test/test_fuzzing_light.ml` which makes intensive use of recursive generators with lists of lists of string, where the perf diff is noticeable, as we knew. Otherwise this looks good.

**Reply by c-cube** [reply]
I don't really know. We have a test suite now which (I think) addresses @jmid 's initial findings, and these bugs were fixed. Y'all at Tezos are using QCheck2 in your test suite without problem? If it's the case I'll try to draft a new release soon.

**Reply by sir4ur0n** [reply]
We currently don't use QCheck2 on `master` because it's not in Opam, but the branch I have linked uses it and has migrated more than 20 PBT files to QCheck2: all tests succeed.

**Reply by sir4ur0n** [reply]
Hi @c-cube , I wanted to kindly bump this discussion :sweat_smile: 

No worries if you don't have time, it's just in case it slipped your mind :smile: 

Is there anything we can do to help get this released?

**Reply by vch9** [reply]
Maybe I could use generators inside https://gitlab.com/tezos/tezos/-/merge_requests/2908 in failing test in order to see the shrinking resilience.

**Comment by c-cube** [comment]
More tests is always good :). I'm aware of this issue but right now I'm on vacation :)

**Comment by c-cube** [comment]
it's released!

**Reply by jmid** [reply]
I can see 0.18 is available here https://opam.ocaml.org/packages/qcheck/
However both https://github.com/c-cube/qcheck and https://c-cube.github.io/qcheck/ still list 0.17 as the latest.
I think the issue is the release tag which is `v0.18` for the latest (in contrast to `0.17`, `0.16`, `0.15`, ...  sans `v`),
which breaks some of the automated updating...

**Reply by c-cube** [reply]
there is no automation for releases, but I just updated the branch :)

**Reply by jmid** [reply]
Hm, OK.
On the frontpage of https://github.com/c-cube/qcheck
in the right column it still says "0.17 (latest)" in my Firefox.

If I then press "Releases (26)" and go to https://github.com/c-cube/qcheck/releases
the "v0.18" is styled differently (smaller, white font) and doesn't say "c-cube released this on ..." like for the other releases 🤔
